File: LENTABP10/src/main/java/com/lenta/lentabp10/models/memory/MemoryTaskProductRepository.java
package com.lenta.lentabp10.models.memory;

import com.lenta.lentabp10.models.repositories.ITaskProductRepository;
import com.lenta.shared.models.core.IProduct;
import com.lenta.shared.models.core.ProductInfo;

import java.util.ArrayList;
import java.util.List;

public class MemoryTaskProductRepository implements ITaskProductRepository {

    protected final List<ProductInfo> productInfo = new ArrayList<>();

    public MemoryTaskProductRepository() {
    }

    @Override
    public ProductInfo findProduct(IProduct product) {
        for(int i=0; i<productInfo.size(); i++) {
            if ( product.getMaterialNumber() == productInfo.get(i).getMaterialNumber() ) {
                return productInfo.get(i);
            }
        }
        return null;
    }

    @Override
    public boolean addProduct(ProductInfo product) {
        if (product == null)
        {
            throw new NullPointerException("product");
        }

        int index = -1;
        for(int i=0; i<productInfo.size(); i++) {
            if ( product.getMaterialNumber() == productInfo.get(i).getMaterialNumber() ) {
                index = i;
            }
        }

        if (index == -1)
        {
            productInfo.add(product);
            return true;
        }

        return false;
    }

    @Override
    public boolean deleteProduct(ProductInfo product) {
        int index = -1;
        for(int i=0; i<productInfo.size(); i++) {
            if ( product.getMaterialNumber() == productInfo.get(i).getMaterialNumber() ) {
                index = i;
            }
        }

        if (index == -1)
        {
            return false;
        }

        productInfo.remove(index);
        return true;
    }

    @Override
    public void clear() {
        productInfo.clear();
    }

    @Override
    public ProductInfo get(int index) {
        return productInfo.get(index);
    }

    @Override
    public int lenght() {
        return productInfo.size();
    }
}

File: writeoff_breakage/src/main/java/com/lenta/bp10/models/memory/MemoryTaskProductRepository.kt
package com.lenta.bp10.models.memory

import com.lenta.bp10.models.repositories.ITaskProductRepository
import com.lenta.shared.models.core.ProductInfo
import java.util.*

class MemoryTaskProductRepository(
        private val products: ArrayList<ProductInfo> = ArrayList()
) : ITaskProductRepository {

    override fun getProducts(): List<ProductInfo> {
        return products.toList()
    }

    override fun findProduct(product: ProductInfo): ProductInfo? {
        return findProduct(product.materialNumber)
    }

    override fun findProduct(materialNumber: String): ProductInfo? {
        return products.firstOrNull { it.materialNumber == materialNumber }
    }

    override fun addProduct(product: ProductInfo): Boolean {
        var index = -1
        for (i in products.indices) {
            if (product.materialNumber == products[i].materialNumber) {
                index = i
            }
        }

        if (index == -1) {
            products.add(product)
            return true
        } else if (index !=  products.size - 1) {
            products.getOrNull(index)?.let {
                products.removeAt(index)
                products.add(it)
            }

        }

        return false
    }

    override fun deleteProduct(product: ProductInfo): Boolean {
        var index = -1
        for (i in products.indices) {
            if (product.materialNumber == products[i].materialNumber) {
                index = i
            }
        }

        if (index == -1) {
            return false
        }

        products.removeAt(index)
        return true
    }

    override fun clear() {
        products.clear()
    }

    override fun get(index: Int): ProductInfo {
        return products[index]
    }

    override fun lenght(): Int {
        return products.size
    }
}